tests/devices/test_default_qubit.py

```diff
@@ -1989,3 +1661,136 @@ def test_s_inverse():
+        assert np.allclose(res, expected_test_output)
+
+    def test_diagonal_operation_case(self, inverse, mocker, monkeypatch):
+        """Tests the case when the operation to be applied is a
+        DiagonalOperation and the _apply_diagonal_unitary method is used."""
+        dev = qml.device('default.qubit', wires=1)
+        par = 0.3
+
+        test_state = np.array([1,0])
+        wires = 0
+        op = qml.PhaseShift(par, wires=wires) if not inverse else qml.PhaseShift(par, wires=wires).inv()
+        assert op.name not in dev._apply_ops
+
+        # Set the internal _apply_diagonal_unitary
+        history = []
+        mock_apply_diag = lambda state, matrix, wires: history.append((state, matrix, wires))
+        with monkeypatch.context() as m:
+            m.setattr(dev, "_apply_diagonal_unitary", mock_apply_diag)
+            assert dev._apply_diagonal_unitary == mock_apply_diag
+
+            dev._apply_operation(test_state, op)
+
+            res_state, res_mat, res_wires = history[0]
+
+            assert np.allclose(res_state, test_state)
+            assert np.allclose(res_mat, np.diag(op.matrix))
+            assert np.allclose(res_wires, wires)
+
+    def test_apply_einsum_case(self, inverse, mocker, monkeypatch):
+        """Tests applying a diagonal operation."""
+        dev = qml.device('default.qubit', wires=1)
+
+        test_state = np.array([1,0])
+        wires = 0
+
+        # Recreate the S gate, so that it's an example for a one-qubit gate
+        # that doesn't inherit from DiagonalOperation
+        class TestSGate(qml.operation.Operation):
+            matrix = np.array([[0, 1], [1, 0]])
+            num_params = 0
+            num_wires = 1
+            par_domain = None
+
+            @classmethod
+            def _matrix(cls, *params):
+                return np.array([[1, 0], [0, 1j]])
+
+        dev.operations.add("TestSGate")
+        op = TestSGate(wires=wires)
+
+        assert op.name in dev.operations
+        assert op.name not in dev._apply_ops
+
+        if inverse:
+            op = op.inv()
+
+        # Set the internal _apply_unitary_einsum
+        history = []
+        mock_apply_einsum = lambda state, matrix, wires: history.append((state, matrix, wires))
+        with monkeypatch.context() as m:
+            m.setattr(dev, "_apply_unitary_einsum", mock_apply_einsum)
+
+            dev._apply_operation(test_state, op)
+
+            res_state, res_mat, res_wires = history[0]
+
+            assert np.allclose(res_state, test_state)
+            assert np.allclose(res_mat, op.matrix)
+            assert np.allclose(res_wires, wires)
+
+    def test_apply_tensordot_case(self, inverse, mocker, monkeypatch):
+        """Tests applying a diagonal operation."""
+        dev = qml.device('default.qubit', wires=3)
+
+        test_state = np.array([1,0])
+        wires = [0, 1, 2]
+
+        # Recreate the Toffoli gate, so that it's an example for a gate with
```

Comment:
```suggestion
        # Redefine the Toffoli gate so that it is an example for a gate with
```

Reply:
I am a bit confused why a gate with two wires would use tensordot?

Reply:
It's due to speed that was observed previously, `_apply_unitary` looks like this:
```python
        if isinstance(operation, DiagonalOperation):
            return self._apply_diagonal_unitary(state, matrix, wires)
        if len(wires) <= 2:
            # Einsum is faster for small gates
            return self._apply_unitary_einsum(state, matrix, wires)

        return self._apply_unitary(state, matrix, wires)
```
This test tests the last `return` statement.